lib/ts/authUtils.ts

```diff
@@ -0,0 +1,613 @@
+                // This means that the session user got primary since we loaded the session user info above
+                sessionUser = await getUser(makeSessionUserPrimaryRes.primaryUserId, userContext);
+
+                if (sessionUser === undefined) {
+                    throw new SessionError({
+                        type: SessionError.UNAUTHORISED,
+                        message: "Session user not found",
+                    });
+                }
+            } else if (makeSessionUserPrimaryRes.status === "OK") {
+                sessionUser = makeSessionUserPrimaryRes.user;
+            } else {
+                // All other statuses signify that we can't make the session user primary
+                // Which means we can't continue
+                return { status: "NON_PRIMARY_SESSION_USER_OTHER_PRIMARY_USER" };
+            }
+        } else {
+            // This means that the app doesn't want to make the session user primary
+            return { status: "NON_PRIMARY_SESSION_USER_NO_LINK" };
+        }
+    }
+    return { status: "OK", sessionUser };
+}
+
+async function checkAuthType( // TODO: better name
+    session: SessionContainerInterface | undefined,
+    accountInfo: AccountInfoWithRecipeId,
+    inputUser: User | undefined,
+    tenantId: string,
+    userContext: UserContext
+): Promise<
+    | { status: "OK"; isFirstFactor: true }
+    | { status: "OK"; isFirstFactor: false; createsNewLinkForTheSessionUser: boolean; sessionUser: User }
+    | { status: "NON_PRIMARY_SESSION_USER_OTHER_PRIMARY_USER" }
+> {
+    logDebugMessage(`checkAuthType called`);
+    let sessionUser: User | undefined = undefined;
+    if (session === undefined) {
+        logDebugMessage(`checkAuthType returning first factor`);
+        // If there is no active session we have nothing to link to - so this has to be a first factor sign in
+        return { status: "OK", isFirstFactor: true };
+    } else {
+        // If the input and the session user are the same
+        if (inputUser !== undefined && inputUser.id === session.getUserId()) {
+            logDebugMessage(`checkAuthType returning secondary factor, session and input user are the same`);
+            // Then this is basically a user logging in with an already linked secondary account
+            // Which is basically a factor completion in MFA terms.
+            // Since the sessionUser and the inputUser are the same in this case, we can just return early
+            return {
+                status: "OK",
+                isFirstFactor: false,
+                createsNewLinkForTheSessionUser: false,
+                sessionUser: inputUser,
+            };
+        }
+        logDebugMessage(`checkAuthType loading session user`);
+        // We have to load the session user in order to get the account linking info
+        const sessionUserResult = await getPrimarySessionUser(session, tenantId, userContext);
+        if (sessionUserResult.status !== "OK") {
```

Comment:
Yeah, I guess we can do that since that also implies that the app tried to make it a primary user.
This'll slightly change how we used to work even in the case where the app doesn't want to link to the session user. I think we are OK with that as well, but that's basically why I went with first factor sign in in this case.